File: code/ex658.cpp
// Binary search & Two pointers

class Solution {
public:
    vector<int> findClosestElements(vector<int>& arr, int k, int x) {
        int n = arr.size();
        
        // Run binary search to find the index of item close to x
        int lo = 0, hi = n - 1, x_index = -1;
        while (lo < hi) {
            int mid = (lo + hi) / 2;
            
            if (arr[mid] == x) {
                x_index = mid;
                break;
            } else if (arr[mid] > x)    hi = mid - 1;
            else    lo = mid + 1;
        }
        
        // Get the index of item which is the closest to x
        int index = -1;
        if (x_index != -1)    index = x_index;
        else {
            if (lo < 0 || lo >= n)    index = hi;
            else if (hi < 0 || hi >= n)    index = lo;
            else if (abs(arr[lo] - x) > abs(arr[hi] - x))    index = hi;
            else    index = lo;
        }
        if (index + 1 < n && abs(arr[index + 1] - x) < abs(arr[index] - x))    index++;
        if (index - 1 >= 0 && abs(arr[index - 1] - x) <= abs(arr[index] - x))    index--;
        
        
        // Two pointers to find interval with k elements which is closest to x
        int p1 = index, p2 = index;
        while (p2 - p1 + 1 < k) {
            if (p1 <= 0)    {p2++;    continue;}
            if (p2 >= n - 1)    {p1--;    continue;}
            
            if (abs(arr[p1 - 1] - x) <= abs(arr[p2 + 1] - x))    p1--;
            else if (abs(arr[p2 + 1] - x) < abs(arr[p1 - 1] - x))    p2++;
        }

        
        // Get the result        
        vector<int> res;
        for (int i = p1; i <= p2; i++) {
            res.push_back(arr[i]);
        }
        
        return res;
    }
};

File: code/ex658.java
class Solution {
    public List<Integer> findClosestElements(int[] arr, int k, int x) {
        int n = arr.length;
        
        // Find the index of element in arr which is closest to x
        int idx = 0;
        if (x <= arr[0])    idx = 0;
        else if (x >= arr[n - 1])    idx = n - 1;
        else {
            int lo = 0, hi = n - 1;
            while (lo <= hi) {
                int mid = lo + (hi - lo) / 2;
                if (arr[mid] == x) {
                    idx = mid;
                    break;
                } else if (arr[mid] < x) {
                    if (arr[mid + 1] >= x) {
                        idx = arr[mid + 1] - x < x - arr[mid] ? mid + 1 : mid;
                        break;
                    }
                    lo = mid + 1;
                } else {
                    if (arr[mid - 1] <= x) {
                        idx = arr[mid] - x < x - arr[mid - 1] ? mid : mid - 1;
                        break;
                    }
                    hi = mid - 1;
                }
            }
        }
        
        // Use two pointers to find k closest elements to x
        int l = idx, r = idx;
        while (r - l + 1 < k) {
            if (l <= 0)    {r++;    continue;}
            if (r >= n - 1)    {l--;    continue;}
            if (x - arr[l - 1] <= arr[r + 1] - x)    l--;
            else    r++;
        }
        
        // Get result
        List<Integer> res = new ArrayList<>();
        for (int i = l; i <= r; i++)
            res.add(arr[i]);
        
        return res;
    }
}